File: atlhyper_web/src/types/overview.ts
/**
 * 集群概览相关类型定义
 */

// 集群健康状态
export interface ClusterHealth {
  status: "Healthy" | "Degraded" | "Unhealthy" | "Unknown";
  reason?: string;
  nodeReadyPercent: number;
  podReadyPercent: number;
}

// 节点就绪统计
export interface NodeReady {
  total: number;
  ready: number;
  percent: number;
}

// 资源使用率
export interface ResourceUsage {
  percent: number;
}

// 概览卡片数据
export interface OverviewCards {
  clusterHealth: ClusterHealth;
  nodeReady: NodeReady;
  cpuUsage: ResourceUsage;
  memUsage: ResourceUsage;
  events24h: number;
}

// 工作负载状态
export interface WorkloadStatus {
  total: number;
  ready: number;
}

// Job 状态
export interface JobStatus {
  total: number;
  running: number;
  succeeded: number;
  failed: number;
}

// 工作负载汇总
export interface WorkloadSummary {
  deployments: WorkloadStatus;
  daemonsets: WorkloadStatus;
  statefulsets: WorkloadStatus;
  jobs: JobStatus;
}

// Pod 状态分布
export interface PodStatusDistribution {
  total: number;
  running: number;
  pending: number;
  failed: number;
  succeeded: number;
  unknown: number;
  runningPercent: number;
  pendingPercent: number;
  failedPercent: number;
  succeededPercent: number;
}

// 峰值统计
export interface PeakStats {
  peakCpu: number;
  peakCpuNode: string;
  peakMem: number;
  peakMemNode: string;
  hasData: boolean;
}

// 工作负载数据
export interface WorkloadsData {
  summary: WorkloadSummary;
  podStatus: PodStatusDistribution;
  peakStats?: PeakStats;
}

// 告警趋势点（按资源类型统计）
export interface AlertTrendPoint {
  at: string;
  kinds: Record<string, number>; // 每种资源类型的告警数量: {"Pod": 5, "Node": 2}
}

// 告警统计
export interface AlertTotals {
  critical: number;
  warning: number;
  info: number;
}

// 最近告警（后端返回小写字段名）
export interface RecentAlert {
  timestamp: string;
  severity: "critical" | "warning" | "info";
  kind: string;
  namespace: string;
  name: string;
  message: string;
  reason: string;
}

// 告警数据
export interface AlertsData {
  trend: AlertTrendPoint[];
  totals: AlertTotals;
  recent: RecentAlert[];
}

// 节点资源使用
export interface NodeUsage {
  node: string;
  cpuUsage: number;
  memUsage: number;
}

// 节点数据
export interface NodesData {
  usage: NodeUsage[];
}

// 集群概览完整数据
export interface ClusterOverview {
  clusterId: string;
  cards: OverviewCards;
  workloads: WorkloadsData;
  alerts: AlertsData;
  nodes: NodesData;
}

// 转换后的概览数据（用于页面展示）
export interface TransformedOverview {
  clusterId: string;
  healthCard: {
    status: string;
    reason: string;
    nodeReadyPct: number;
    podHealthyPct: number;
  };
  nodesCard: {
    totalNodes: number;
    readyNodes: number;
    nodeReadyPct: number;
  };
  cpuCard: {
    percent: number;
  };
  memCard: {
    percent: number;
  };
  alertsTotal: number;
  // 工作负载统计
  workloads: {
    deployments: { total: number; ready: number };
    daemonsets: { total: number; ready: number };
    statefulsets: { total: number; ready: number };
    jobs: { total: number; running: number; succeeded: number; failed: number };
  };
  podStatus: {
    total: number;
    running: number;
    pending: number;
    failed: number;
    succeeded: number;
    runningPercent: number;
    pendingPercent: number;
    failedPercent: number;
    succeededPercent: number;
  };
  peakStats: {
    peakCpu: number;
    peakCpuNode: string;
    peakMem: number;
    peakMemNode: string;
    hasData: boolean;
  };
  alertTrends: {
    ts: number;
    kinds: Record<string, number>; // 按资源类型统计
  }[];
  recentAlerts: {
    time: string;
    severity: string;
    kind: string;
    namespace: string;
    message: string;
    reason: string;
    name: string;
  }[];
  nodeUsages: {
    nodeName: string;
    cpuPercent: number;
    memoryPercent: number;
  }[];
}

File: model_v3/cluster/overview.go
package cluster

import "time"

// ClusterOverview 集群概览
type ClusterOverview struct {
	ClusterID string            `json:"clusterId"`
	Cards     OverviewCards     `json:"cards"`
	Workloads OverviewWorkloads `json:"workloads"`
	Alerts    OverviewAlerts    `json:"alerts"`
	Nodes     OverviewNodes     `json:"nodes"`
}

type OverviewCards struct {
	ClusterHealth ClusterHealth   `json:"clusterHealth"`
	NodeReady     ResourceReady   `json:"nodeReady"`
	CPUUsage      ResourcePercent `json:"cpuUsage"`
	MemUsage      ResourcePercent `json:"memUsage"`
	Events24h     int             `json:"events24h"`
}

type ClusterHealth struct {
	Status           string  `json:"status"`
	Reason           string  `json:"reason,omitempty"`
	NodeReadyPercent float64 `json:"nodeReadyPercent"`
	PodReadyPercent  float64 `json:"podReadyPercent"`
}

type ResourceReady struct {
	Total   int     `json:"total"`
	Ready   int     `json:"ready"`
	Percent float64 `json:"percent"`
}

type ResourcePercent struct {
	Percent float64 `json:"percent"`
}

type OverviewWorkloads struct {
	Summary   WorkloadSummary       `json:"summary"`
	PodStatus PodStatusDistribution `json:"podStatus"`
	PeakStats *PeakStats            `json:"peakStats,omitempty"`
}

type WorkloadSummary struct {
	Deployments  WorkloadStatus `json:"deployments"`
	DaemonSets   WorkloadStatus `json:"daemonSets"`
	StatefulSets WorkloadStatus `json:"statefulSets"`
	Jobs         JobStatus      `json:"jobs"`
}

type WorkloadStatus struct {
	Total int `json:"total"`
	Ready int `json:"ready"`
}

type JobStatus struct {
	Total     int `json:"total"`
	Running   int `json:"running"`
	Succeeded int `json:"succeeded"`
	Failed    int `json:"failed"`
}

type PodStatusDistribution struct {
	Total            int     `json:"total"`
	Running          int     `json:"running"`
	Pending          int     `json:"pending"`
	Failed           int     `json:"failed"`
	Succeeded        int     `json:"succeeded"`
	Unknown          int     `json:"unknown"`
	RunningPercent   float64 `json:"runningPercent"`
	PendingPercent   float64 `json:"pendingPercent"`
	FailedPercent    float64 `json:"failedPercent"`
	SucceededPercent float64 `json:"succeededPercent"`
}

type PeakStats struct {
	PeakCPU     float64 `json:"peakCpu"`
	PeakCPUNode string  `json:"peakCpuNode"`
	PeakMem     float64 `json:"peakMem"`
	PeakMemNode string  `json:"peakMemNode"`
	HasData     bool    `json:"hasData"`
}

type OverviewAlerts struct {
	Trend  []AlertTrendPoint `json:"trend"`
	Totals AlertTotals       `json:"totals"`
	Recent []RecentAlert     `json:"recent"`
}

type AlertTrendPoint struct {
	At    time.Time      `json:"at"`
	Kinds map[string]int `json:"kinds"`
}

type AlertTotals struct {
	Critical int `json:"critical"`
	Warning  int `json:"warning"`
	Info     int `json:"info"`
}

type RecentAlert struct {
	Timestamp string `json:"timestamp"`
	Severity  string `json:"severity"`
	Kind      string `json:"kind"`
	Namespace string `json:"namespace"`
	Name      string `json:"name"`
	Message   string `json:"message"`
	Reason    string `json:"reason"`
}

type OverviewNodes struct {
	Usage []NodeUsage `json:"usage"`
}

type NodeUsage struct {
	Node     string  `json:"node"`
	CPUUsage float64 `json:"cpuUsage"`
	MemUsage float64 `json:"memUsage"`
}

func CalculateHealthStatus(nodeReadyPct, podReadyPct float64) string {
	if nodeReadyPct >= 90 && podReadyPct >= 80 {
		return "Healthy"
	}
	if nodeReadyPct >= 70 && podReadyPct >= 60 {
		return "Degraded"
	}
	return "Unhealthy"
}

func CalculateHealthReason(nodeReadyPct, podReadyPct float64) string {
	if nodeReadyPct < 70 {
		return "节点就绪率过低"
	}
	if podReadyPct < 60 {
		return "Pod 就绪率过低"
	}
	if nodeReadyPct < 90 {
		return "部分节点未就绪"
	}
	if podReadyPct < 80 {
		return "部分 Pod 未就绪"
	}
	return ""
}
